# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webauthn/WebAuthnResult.h
# Commit: 8b0c234c92ad
# Full Hash: 8b0c234c92ad8db923d5150aba9e7385e8f5c33c
# Author: John Schanck <jschanck@mozilla.com>
# Date: 2024-12-05 09:58:04
# Regressor Bug: 1935278
# File Overlap Count: 1
# Description:
#   Bug 1935278 - support for the WebAuthn PRF extension on Windows. r=keeler
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D230316
# ==============================================================================

diff -r de1a42417976 -r 8b0c234c92ad dom/webauthn/WebAuthnResult.h
--- a/dom/webauthn/WebAuthnResult.h	Wed Dec 04 23:59:18 2024 +0000
+++ b/dom/webauthn/WebAuthnResult.h	Wed Dec 04 23:59:19 2024 +0000
@@ -96,6 +96,7 @@
                 (BOOL*)pExtension->pvExtension;
             if (*pCredentialCreatedWithHmacSecret) {
               mHmacCreateSecret = Some(true);
+              mPrf = Some(true);
             }
           }
         }
@@ -132,6 +133,12 @@
         mAuthenticatorAttachment = Some(u"cross-platform"_ns);
       }
     }
+
+    if (aResponse->dwVersion >= WEBAUTHN_CREDENTIAL_ATTESTATION_VERSION_5) {
+      if (aResponse->bPrfEnabled) {
+        mPrf = Some(true);
+      }
+    }
   }
 #endif
 
@@ -144,6 +151,7 @@
   Maybe<nsCString> mClientDataJSON;
   Maybe<bool> mCredPropsRk;
   Maybe<bool> mHmacCreateSecret;
+  Maybe<bool> mPrf;
   Maybe<nsString> mAuthenticatorAttachment;
 };
 
@@ -210,6 +218,18 @@
                                       aResponse->cbAuthenticatorData);
 
     mAuthenticatorAttachment = Nothing();  // not available
+    if (aResponse->pHmacSecret) {
+      if (aResponse->pHmacSecret->cbFirst > 0) {
+        mPrfFirst.emplace();
+        mPrfFirst->AppendElements(aResponse->pHmacSecret->pbFirst,
+                                  aResponse->pHmacSecret->cbFirst);
+      }
+      if (aResponse->pHmacSecret->cbSecond > 0) {
+        mPrfSecond.emplace();
+        mPrfSecond->AppendElements(aResponse->pHmacSecret->pbSecond,
+                                   aResponse->pHmacSecret->cbSecond);
+      }
+    }
   }
 #endif
 
@@ -223,6 +243,8 @@
   nsTArray<uint8_t> mUserHandle;
   Maybe<nsString> mAuthenticatorAttachment;
   Maybe<bool> mUsedAppId;
+  Maybe<nsTArray<uint8_t>> mPrfFirst;
+  Maybe<nsTArray<uint8_t>> mPrfSecond;
 };
 
 }  // namespace mozilla::dom